### ISSUE TITLE
Allow enabling extensions in a settings file

### ISSUE THREAD

**Issue by wolftune** [ndmitchell/hlint#223]
I really don't like continually seeing parse errors for things like `[whamlet|` blocks. I get that hlint can't deal with whamlet. It should ignore it and similar cases instead of reporting the parse error.

**Comment by ndmitchell** [comment]
It's not a parse error for there to be arbitrary stuff in a whamlet block, and HLint is meant to ignore it. Taking the first example at http://www.yesodweb.com/book/basics#basics_hello_world I am able to check that module with no errors. Can you give me an example of a whamlet block that doesn't work?

**Comment by wolftune** [comment]
src/Handler/NewDesign.hs in [Snowdrift](https://github.com/snowdriftcoop/snowdrift) has a parse error on the h3 line in:

```
projectNav :: ProjectHandle -> Widget
projectNav handle =
    [whamlet|
        <h3>Subpages
        <ul>
            <li><a href=@{PUpdatesR handle}>Updates
            <li><a href=@{WikiPagesR handle}>Wiki</a> (links to pre-alpha)
            <li><a href=@{ProjectDiscussionR handle}>Discussion</a> (links to pre-alpha)
            <li><a href=@{PTransactionsR handle}>Transactions
    |]
```

also src/Handler/User/Create.hs on the line with form

```
        [whamlet|
            <form method=POST>
                ^{form}
                <input type=submit>
        |]
```

Also src/View/Time.hs:19:9

```
        toWidget [hamlet|
  >         <span title=#{render time}>
                #{showDiffTime now time}&nbsp;ago
        |]
```

and other cases

**Comment by ndmitchell** [comment]
The error is because you don't have `{-# LANGUAGE QuasiQuotes #-}` at the top of your file, and instead use it as a default-extension in your .cabal file. HLint turns on many extensions, but QuasiQuotes is one that has a habit of breaking modules not designed for it, so I explicitly turn it off. You can solve that by adding `-XQuasiQuotes` on the command line when running HLint, which will tell HLint what you have already told Cabal about default extensions.

**Comment by wolftune** [comment]
Is there a way to add this to an hlint settings file or something? As hlint is integrated with other tools and I also want to minimize the room for manual human command adjustment, it would be much better if there were a way to just set HLint to use this option by default…

Incidentally, would you recommend us setting that language extension per file anyway/instead?

**Comment by ndmitchell** [comment]
I generally use all extensions explicitly per file, but its really up to you - different people have different preferences, which is why HLint tries to turn on all the extensions it think it can do without breaking stuff. It's just unfortunate in your case the one you need does break stuff sometimes.

There's no way to add it to a settings file yet, but I'll leave this bug open to do that.

**Comment by Cypher1** [comment]
Would really like a .hlintrc like pylint has

**Comment by lwm** [comment]
@ndmitchell Can you provide some hints as to what a `.hlintrc` file might look like? I'd like to take a stab at it. Would something like [ini](https://github.com/chrisdone/ini) be sufficient? Something like:

```
[EXTENSIONS]
quasiquotes=True
```

Not sure.

**Comment by ndmitchell** [comment]
I would imagine you could put extensions in the `Hlint.hs` file with:

```
{-# LANGUAGE QuasiQuotes #-}
```

Furthermore it seems to make sense to allow arbitrary command line flags in the `Hlint.hs`, for example:

```
{-# OPTIONS_HLINT --ignore=foo #-}
```

**Comment by ndmitchell** [comment]
You can now create a `.hlint.yaml` file and add `- arguments: [-XQuasiQuote]`

**Comment by lwm** [comment]
:tada:

**Comment by ndmitchell** [comment]
Btw, I've now got the release almost ready to go, so if anyone wants to try it out I'd welcome any feedback before the actual release.

**Comment by wolftune** [comment]
does the .hlint.yaml file go universally in ~/ or in a project's directory or either/both…?

**Comment by ndmitchell** [comment]
@wolftune we search the current directory upwards, so a projects directory. I considered in ~/, but in the end decided to replicate the stylish-haskell logic, since I'm trying to be more consistent with other tools.

**Comment by wolftune** [comment]
thanks

**Comment by AlexeyRaga** [comment]
@ndmitchell should it also work for `MultiParamTypeClasses`?
It is in my `default-extensions`, but `hlint` still wants me to add it to each file.

I tried adding 
```
- arguments: [-XMultiParamTypeClasses]
```

But it didn't seem to work.

**Comment by ndmitchell** [comment]
@AlexeyRaga That works for me to enable extensions, however, you don't need to enable MultiParamTypeClasses as that's on by default, so it's no great surprise enabling it doesn't have any impact. What error are you seeing or change are you trying to get?

**Comment by AlexeyRaga** [comment]
@ndmitchell Oh, sorry, I think the error is actually from `stylish-haskell` and not from `hlint`!
Sorry for making noise!

**Comment by vyorkin** [comment]
hm, it doesn't seem to work when I try to enable the `-XUnicodeSyntax` extension.
my example file:
```haskell
module Model.File where

import qualified Data.Bson as Bson
import Data.Int (Int64)
import Data.Time.Clock (UTCTime)
import GHC.Generics (Generic)

-- | Represents a file stored in Mongo GridFS.
data File = File
  { _id        ∷ Bson.ObjectId
  , length     ∷ Int64
  , uploadDate ∷ UTCTime
  , fileName   ∷ String
  } deriving (Eq, Show, Generic)
```

```bash
 ▲ haskell/2018/spreadsheet-api cat .hlint.yaml
- arguments: [-XUnicodeSyntax]

 ▲ haskell/2018/spreadsheet-api hlint src/Model/File.hs --refactor
refactor: (RealSrcSpan SrcSpanOneLine "src/Model/File.hs" 10 16 17,"parse error on input \8216\8759\8217")
CallStack (from HasCallStack):
  error, called at src/Main.hs:189:22 in main:Main
```

**Comment by ndmitchell** [comment]
@vyorkin Does it work if you don't pass `--refactor`? My guess is that `-XUnicodeSyntax` isn't required either in the `.hlint.yaml` or on the command line. And separately refactor can't deal with unicode, perhaps unless you pass some argument. I'm going to raise a separate bug for that.

**Comment by ndmitchell** [comment]
I opened https://github.com/ndmitchell/hlint/issues/465 to track the refactor issue separately. I don't think it has anything to do with extensions in a settings file.

**Comment by vyorkin** [comment]
Without `--refactor` it just outputs `no hints` (as it should be, I guess).
But if I introduce some issue (like duplicate import)
```
...
import GHC.Generics (Generic)
import GHC.Generics (Generic)
...
```

it correctly reports it:
```
▲ play/haskell/tmp hlint File.hs                                                                                                                                                                                                                                           2m
File.hs:6:1: Warning: Use fewer imports
Found:
  import GHC.Generics (Generic)
  import GHC.Generics (Generic)

Why not:
  import GHC.Generics (Generic)


1 hint

 △ play/haskell/tmp hlint --refactor File.hs
refactor: (RealSrcSpan SrcSpanOneLine "File.hs" 11 16 17,"parse error on input \8216\8759\8217")
CallStack (from HasCallStack):
  error, called at src/Main.hs:189:22 in main:Main
```

**Comment by ndmitchell** [comment]
Yep, since it doesn't touch refactor unless there are hints to be dealt with. Leaving this to #465

**Comment by Martinsos** [comment]
> You can now create a `.hlint.yaml` file and add `- arguments: [-XQuasiQuote]`

@ndmitchell This answer was exactly what I was looking for, but there is typo: it should be QuasiQuote**s**, now QuasiQuote (s is missing!). It took me some time to figure out that was the problem!

**Comment by ndmitchell** [comment]
Thanks for mentioning that @Martinsos - I've raised #1093 to emit a warning - way too easy to make a typo here

**Comment by Martinsos** [comment]
> Thanks for mentioning that @Martinsos - I've raised #1093 to emit a warning - way too easy to make a typo here

Great, that is even better :)!
